### DISCUSSION TITLE
Calculating timestamps

### DISCUSSION THREAD

**Discussion by sandrojologua** [CoderLine/alphaTab#1569] [General]
Hello there, what is the best way to calculate timestamps of notes/beats using alphatab library? I see that alphatab uses midi to retrieve timestamps, but I find it little bit hard to understand. Is there any fast and efficient way to do this calculation only? with or without midi?
Thanks.

**Comment by Danielku15** [comment]
Yes, alphaTab is handling internally things in Midi ticks and translating it to real timestamps requires some calculations which depend on the current tempo (BPMs). 

If you explain a bit more what you want to build, then I can provide you already some snippet/example how to achieve your goal.  

The topic can be tricky because of following reasons:

* Songs can have repeats and repeat endings having a big influence when a beat might be played the first time. 
* You can change the playback speed (e.g. speed trainer) which change the millisecond position. 
* Midi ticks are a tempo (BPM) independent time axis. To translate ticks to milliseconds involves looking at the tempo changes happening during the song. 

https://github.com/CoderLine/alphaTab/issues/1580

Some hints: 

* During playback you get the information in ticks and milliseconds at any time: 
https://www.alphatab.net/docs/reference/api/playerpositionchanged/
* The MidiTickLookup (low level apis) is a component filled when we translate the song into midi and it holds more accurate midi tick information with repeats and tempo changes respected.https://www.alphatab.net/docs/reference/api/tickcache#miditicklookupfindbeat-signatures
*  Internally in alphaTab there are functions to translate this info. https://github.com/CoderLine/alphaTab/blob/d28966db841b4c6c1376b5360d770d782ac12960/src/synth/MidiFileSequencer.ts#L292

In Guitar Pro there is a feature to show timestamps above bars. But we don't have this on the feature roadmap yet. 
![image](https://github.com/user-attachments/assets/e58a70b6-46c7-4db1-9b1f-6bfd347e864a)
![image](https://github.com/user-attachments/assets/d69104d5-9301-48bc-8b91-7462427aea19)
